### DISCUSSION TITLE
RAG with .NET: Session Resources

### DISCUSSION THREAD

**Discussion by multispark** [microsoft/RAG_Hack#25] [Announcements]
This is the resources and discussion thread for this live stream: https://aka.ms/raghack/dotnet
It will be updated with links after the session.

Recording: https://www.youtube.com/watch?v=Rkw-N65Kfww
Slides: https://github.com/microsoft/RAG_Hack/blob/main/slides/2024%2009%2003%20RAGHack%20NET.pptx

Please ask any follow-up questions here, and the speakers will see them.

**Comment by elbruno** [comment]
Hi, 
sharing some of the links that we used during the session:

- https://aka.ms/ragchatnet
- https://aka.ms/phi-3CookBook
- https://github.com/dotnet/eShopSupport

Best,
Bruno

**Comment by negar-mir** [comment]
Hi @elbruno,
I have an Azure OpenAI instance with uploaded data (Bring Your Own Data - BYOD) in my project. I am using the Azure OpenAI client library for .NET (Azure.AI.OpenAI version 2.0.0-beta.4). The chat functionality with the instance works well, and it retrieves answers composed from the uploaded data.

Additionally, I can retrieve dynamic data from an external API using tools and tools_choice, which functions as expected. However, when I try to combine these features, the tools call is ignored, and only the uploaded data is used to provide answers.

I found a note in the Azure OpenAI client library documentation that states I cannot use tools calls and BYOD together in a single request. It suggests separating the evaluation of Chat Functions and Azure Chat Extensions across multiple requests in my solution design.

I appreciate any guidance you can provide in advance, as I am a bit confused about how to implement this separation. Is there any sample code or documentation I can follow for assistance with this?

**Reply by elbruno** [reply]
Hi @negar-mir 

I wasn't aware about that limitation of the Azure OpenAI library. Have you tried using Semantic Kernel? That is the scenario we showcased and it works with function calling and Azure OpenAI Models

Best
